text/3503-frontmatter.md

```diff
@@ -0,0 +1,486 @@
+
+## Syntax
+
+When choosing a syntax for embedding manifests in source files, our care-abouts are
+- How obvious it is for new users when they see it
+- How easy it is for newer users to remember it and type it out
+- How machine editable it is for `cargo add` and friends
+- Needs to be valid Rust code for quality of error messages, etc
+- Leave the door open in case we want to reuse the syntax for embedded lockfiles
+- Leave the door open for single-file `lib`s
+
+Benefits for the proposed solution are:
+- Visually/syntactically lightweight
+- Has parallels to ideas outside of Rust, building on external knowledge that might exist
+- Easy for cargo to parse and modify
+  - Cargo needs a simple syntax subset to parse to ensure it can adequately report errors for the part it cares about but does not get in the way of rustc reporting the high quality errors it cares about
+- In the future, this can be leveraged by other build systems or tools
+
+### Alternative 1: Doc-comment
+
+```rust
+#!/usr/bin/env cargo
+
+//! ```cargo
+//! [package]
+//! edition = "2018"
+//! ```
+
+fn main() {
+}
+```
+
+Benefits
+- Parsers are available to make this work (e.g. `syn`, `pulldown-cmark`)
+- Familiar syntax both to read and write.
+  - When discussing with a Rust author, it was pointed out many times people preface code with a comment specifying the dependencies ([example](https://github.com/dtolnay/prettyplease#example)), this is the same idea but reusable by cargo
+  - When discussing on forums, people expressed how they had never seen the syntax but instantly were able to understand it
+
+Downsides:
+- When discussing with a a Rust crash course teacher, it was felt their students would have a hard time learning to write these manifests from scratch
+  - Having the explain the overloading of concepts to new users
+  - Unpredictable location (both the doc comment and the cargo code block within it)
+  - Visual clutter (where clutter is overwhelming already in Rust)
+- Might be a bit complicated to do edits (translating between location within
+  `toml_edit` spans to the location within `syn` spans especially with different comment styles)
+- Either we expose `syn`s lesser parse errors or we skip errors, deferring to rustc's, but then have the wrong behavior on commands that don't invoke rustc, like `cargo metadata`
+- Requires pulling in a full markdown parser to extract the manifest
+  - Incorrectly formatted markdown would lead to a missing manifest and confusing error messages at best or silent incorrect behavior at worse
+
+### Alternative 2: Macro
+
+```rust
+#!/usr/bin/env cargo
+
+cargo! {
+[package]
+edition = "2018"
+}
+
+fn main() {
+}
+```
+Benefits
+- Parsers are available to make this work (e.g. `syn`)
+
+Downsides
+- The `cargo` macro would need to come from somewhere (`std`?) which means it is taking on `cargo`-specific knowledge
```

Comment:
For the macro approach, I don't think it would be necessary to embed any Cargo-specific knowledge in std. In every other approach the data here is stored in a glorified comment, which means we're fine if it gets thrown away as far as Rust is concerned. The macro here could simply expand to nothing, and trust that other tooling will parse the macro body as needed (which is easier than it sounds, since the macro body should just be treated as raw tokens rather than anything that needs parsing). Rather than calling it cargo, call it `build!` or `meta!` or something.

Reply:
I added mention of `meta!` in 42077b7f.  I don't bother exploring it due to the other problems with macros.